File: src/java/graphql/spring-graphql/src/main/java/benchmark/QueryController.java
package benchmark;

import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.stereotype.Controller;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Controller
public class QueryController {

    private final DatabaseService databaseService;
    private final CacheService cacheService;

    public QueryController(DatabaseService databaseService, CacheService cacheService) {
        this.databaseService = databaseService;
        this.cacheService = cacheService;
    }

    @QueryMapping
    public Models.Health health() {
        return new Models.Health(
                "ok",
                "1.0.0",
                Instant.now().toString(),
                databaseService.checkHealth(),
                cacheService.checkHealth()
        );
    }

    @QueryMapping
    public Models.JsonItemsResult jsonItems(@Argument int limit) {
        int count = Canonical.itemCount(limit);
        List<Models.JsonItem> items = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            items.add(new Models.JsonItem(
                    i,
                    Canonical.uuid(i),
                    Canonical.name(i),
                    Canonical.email(i),
                    Canonical.CREATED_AT,
                    Canonical.isActive(i)
            ));
        }
        return new Models.JsonItemsResult(items, items.size(), Instant.now().toString());
    }

    @QueryMapping
    public Models.User user(@Argument int id) {
        return databaseService.getUser(id);
    }

    @QueryMapping
    public Models.ComplexOrdersResult complexOrders(@Argument int days) {
        return databaseService.getComplexOrders(days);
    }

    @QueryMapping
    public Models.CacheEntry cache(@Argument String key) {
        return cacheService.getCacheEntry(key);
    }
}

File: src/kotlin/graphql/spring-graphql/src/main/kotlin/benchmark/QueryController.kt
package benchmark

import org.springframework.graphql.data.method.annotation.Argument
import org.springframework.graphql.data.method.annotation.QueryMapping
import org.springframework.stereotype.Controller
import java.util.UUID

@Controller
class QueryController(
    private val databaseService: DatabaseService,
    private val cacheService: CacheService
) {

    @QueryMapping
    fun health(): Health {
        val dbOk = databaseService.checkHealth()
        val cacheOk = cacheService.checkHealth()
        return Health(
            status = if (dbOk && cacheOk) "healthy" else "unhealthy",
            version = "1.0.0",
            timestamp = java.time.Instant.now().toString(),
            database = if (dbOk) "connected" else "disconnected",
            cache = if (cacheOk) "connected" else "disconnected"
        )
    }

    @QueryMapping
    fun jsonItems(@Argument limit: Int): JsonItemsResult {
        val timestamp = java.time.Instant.now().toString()
        val items = (1..limit).map { i ->
            JsonItem(
                id = i,
                uuid = UUID.randomUUID().toString(),
                name = "Item $i",
                email = "user${i}@example.com",
                createdAt = timestamp,
                isActive = i % 2 == 0
            )
        }
        return JsonItemsResult(
            items = items,
            count = limit,
            timestamp = timestamp
        )
    }

    @QueryMapping
    fun user(@Argument id: Int): User? {
        return databaseService.getUser(id)
    }

    @QueryMapping
    fun complexOrders(@Argument days: Int): ComplexOrdersResult {
        val data = databaseService.getComplexOrders(days)
        return ComplexOrdersResult(
            periodDays = days,
            totalUsers = data.size,
            data = data
        )
    }

    @QueryMapping
    fun cache(@Argument key: String): CacheEntry {
        val cached = cacheService.get(key)
        if (cached != null) {
            return CacheEntry(
                key = key,
                value = cached,
                cached = true,
                ttl = 300
            )
        }
        val value = "Cache value for $key at ${java.time.Instant.now()}"
        cacheService.set(key, value, 300)
        return CacheEntry(
            key = key,
            value = value,
            cached = false,
            ttl = 300
        )
    }
}
